src/Chapter4.hs

```diff
@@ -628,8 +661,14 @@ Can you implement a monad version of AND, polymorphic over any monad?
 
 🕯 HINT: Use "(>>=)", "pure" and anonymous function
 -}
+andMv1 :: (Monad m) => m Bool -> m Bool -> m Bool
+andMv1 a b = fmap (&&) a <*> b
+
 andM :: (Monad m) => m Bool -> m Bool -> m Bool
-andM = error "andM: Not implemented!"
+andM a b = a >>= \x -> pure (x &&) <*> b
+
+andMWithListProblem :: (Monad m) => m Bool -> m Bool -> m Bool
+andMWithListProblem a b = a >>= \x -> if x then b else (pure False)
```

Comment:
Answering your question regarding the behaviour. No, it's not a bug :slightly_smiling_face: A lazier version stops earlier. If you want to match all elements of lists pairwise, you need to use one of the previous implementations of `andM` that you've provided.

Basically, these are just different functions with different behaviours and for your real problem you should use the one that works for your best. However, I want to add that lists are rarely used as monads, even though they have a `Monad` instance. The monadic behaviour for lists has only very few use-cases, unlike data types like `Maybe` or `Either`. So most generic functions are usually optimized for these types, that's why you will find this lazy version of `andM` in most libraries out there.

Reply:
Oh, I see! Thank you for the clarification! 😊